Created EntityPersistenceNamingStrategy for replacing FileNamingStrategy

ahamshubham opened the pull request:
Resolves #246. Changes made:

- Added deprecated tag to javadoc of `FileNamingStrategy.java`
- Created `EntityPersistenceNamingStrategy.java` as a copy of `FileNamingStrategy.java` for replacing `FileNamingStrategy.java`
- Refactored the code by changing the order of the functions for better readability

ahamshubham commented:
!test

ahamshubham commented:
- Replaced occurences of `FileNamingStrategy` with `EntityPersistenceNamingStrategy`
- Updated the javadocs to reflect the changes in the file name
- Moved `HierarchicFileNamingStrategy.java` and `HierarchicFileNamingStrategyTest.groovy` to package `edu.ie3.datamodel.io.naming`

ahamshubham commented:
Sonarqube shows several warnings with respect to the regex patterns declared in `EntityPersistenceNamingStrategy.java`: 'Use the named groups of this regex or remove the names.'. Also, some Codacy Static Code Analysis issues were with respect to the size of some functions in `EntityPersistenceNamingStrategyTest.groovy`. Should I work on resolving these issues?

ckittl commented:
> Sonarqube shows several warnings with respect to the regex patterns declared in `EntityPersistenceNamingStrategy.java`: 'Use the named groups of this regex or remove the names.'. Also, some Codacy Static Code Analysis issues were with respect to the size of some functions in `EntityPersistenceNamingStrategyTest.groovy`. Should I work on resolving these issues?

The code quality of tests in detail doesn't really matter to me. But if there is something in production code, we should try to fix it (cf. "camp site rule" or "scout's rule" of clean code).

ahamshubham commented:
I have fixed the errors pointed out by Codacy and I tried fixing the issue pointed out by Sonarqube which says that there are unused named groups in the regex patterns in the file 'EntityPersistenceNamingStrategy.java'. To resolve the issue, I removed the function call to get the member variable with the member variable itself (since the call was being made inside the class of which the variable is a member variable). The issue still exists according to Sonarqube. The problem is that the named groups are actually being used in the same file (contradictory to what Sonarqube says) and so cannot be removed as Sonarqube points out.

ahamshubham commented:
Quality gate test, however, is failing because I guess there is duplicate code as we have both `FileNamingStrategy.java` and `EntityPersistenceNamingStrategy.java` and their respective test case files in the code.

ahamshubham commented:
This commit can be reverted if the files are not supposed to be deleted.

ckittl commented on src/test/groovy/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategyTest1.groovy:
```diff
@@ -0,0 +1,529 @@
+import java.time.ZonedDateTime
+import java.util.regex.Pattern
+
+class EntityPersistenceNamingStrategyTest1 extends Specification {
```
Why did you split up the test tasks? If it still is sensible, please give it a reasonable naming.

ahamshubham replied:
Actually, the number of functions in `EntityPersistenceNamingStrategyTest.groovy` was more than 30 due to which Codacy raised an issue. I decided to split the test cases into two files to fix the issue but I could not come up with some way in which I can decide the test case files to have a meaningful name. I will merge them both together again.

ckittl reviewed (changes requested):
Looks good by around 90 % to me. Please try and improve the comments, they are mainly regarding documentation issues.

ahamshubham commented on docs/readthedocs/io/csvfiles.rst:
```diff
@@ -4,25 +4,26 @@ csv files
-   A basic file naming strategy that is able to add prefix and suffix to the file names. A flat folder structure is
-   considered. For more details see `Default naming strategy`_.
+1. **EntityPersistenceNamingStrategy**:
+   A basic naming strategy that is able to add prefix and suffix to the names of the data sinks. A flat folder structure
```
Will make the change

ahamshubham commented on docs/readthedocs/io/csvfiles.rst:
```diff
@@ -4,25 +4,26 @@ csv files
 Default naming strategy
 =======================
-There is a default mapping from model class to file naming in the case you would like to use csv files for
+There is a default mapping from model class to naming of data sinks in the case you would like to use csv files for
```
Will make the change

ahamshubham commented on docs/readthedocs/io/csvfiles.rst:
```diff
@@ -4,25 +4,26 @@ csv files
+There is a default mapping from model class to naming of data sinks in the case you would like to use csv files for
 (de-)serialization of models.
-You may extend / alter the naming with pre- or suffix by calling :code:`new FileNamingStrategy("prefix", "suffix")`.
+You may extend / alter the naming with pre- or suffix by calling :code:
+`new EntityPersistenceNamingStrategy("prefix","suffix")`.
```
Will make the change.

ahamshubham commented on src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:
```diff
@@ -0,0 +1,506 @@
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Provides an easy to use standard way to name files, tables or any other data sink that is used in
+ * order to persist entities. Normal use cases are e.g., I/O operations with .csv files or
+ * databases. If a folder structure is required for file based I/O operations, one might consider
+ * using {@link HierarchicFileNamingStrategy}
```
Will make the change.

ahamshubham commented on src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:
```diff
@@ -0,0 +1,506 @@
+   * @param cls Targeted class of the given file
+   * @return An optional sub path to the actual file
+   */
+  public Optional<String> getFilePath(Class<? extends UniqueEntity> cls) {
```
Okay, so I do not make any changes for this then.

ahamshubham commented on src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:
```diff
@@ -0,0 +1,506 @@
+   * @param typeClass the asset type class a filename string should be generated from
+   * @return the filename string
+   */
+  public Optional<String> getTypeFileName(Class<? extends AssetTypeInput> typeClass) {
```
Will make the change.

ahamshubham commented on src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:
```diff
@@ -0,0 +1,506 @@
+   * @return The file name
+   */
+  public Optional<String> getFileName(Class<? extends UniqueEntity> cls) {
+    Optional<String> inputEntityFileName = getInputEntityFileName(cls);
+    if (inputEntityFileName.isPresent()) return inputEntityFileName;
+    if (ResultEntity.class.isAssignableFrom(cls))
+      return getResultEntityFileName(cls.asSubclass(ResultEntity.class));
+    logger.error("There is no naming strategy defined for {}", cls.getSimpleName());
+    return Optional.empty();
```
Thank you. Will make the change.

ckittl replied:
Thanks for pointing at this. Actually, the instances of `CharacteristicInput` have no need to be persisted anymore. Therefore, I deleted the parts, that caused the issues.

ahamshubham commented on src/main/java/edu/ie3/datamodel/io/naming/EntityPersistenceNamingStrategy.java:
```diff
@@ -0,0 +1,506 @@
+   * @param resultEntityClass the result entity class a filename string should be generated from
+   * @return the filename string
+   */
+  public Optional<String> getResultEntityFileName(Class<? extends ResultEntity> resultEntityClass) {
```
Will make the change.
